src/lib/Endpoint/SqlSrvUsers.php

```diff
@@ -0,0 +1,690 @@
+                    $this->addRoles($sqlName, $object[self::ATTRUSERROLES], $simulate);
+                }
+            } catch (InvalidQuery $e) {
+                $this->logger->error('failed to create new sql user', [
+                    'class' => get_class($this),
+                    'exception' => $e,
+                ]);
+
+                return null;
+            }
+        } else {
+            $this->logger->info('no sql-user has been created', [
+                'category' => get_class($this),
+            ]);
+        }
+
+        return (string) $principalId;
+    }
+
+    public function delete(AttributeMapInterface $map, array $object, EndpointObjectInterface $endpoint_object, bool $simulate = false): bool
+    {
+        $sqlUserQuery = '';
+        $endpoint_object = $endpoint_object->getData();
+
+        if (isset($endpoint_object[self::ATTRSQLNAME])) {
+            $sqlUserQuery = $this->dropSqlUserQuery($endpoint_object[self::ATTRSQLNAME]);
+        }
+
+        $loginQuery = $this->dropLoginQuery($endpoint_object[self::ATTRLOGINNAME]);
+
+        try {
+            if ($sqlUserQuery !== '') {
+                $this->socket->query($sqlUserQuery, $simulate);
+            }
+            $this->socket->query($loginQuery, $simulate);
+
+            return true;
+        } catch (InvalidQuery $e) {
+            $this->logger->error('failed to delete object with query', [
+                'class' => get_class($this),
+                'exception' => $e,
+            ]);
+
+            return false;
+        }
+    }
+
+    public function getDiff(AttributeMapInterface $map, array $diff): array
+    {
+        $result = [];
+        foreach (self::UPDATEATTRIBUTES as $attr) {
+            if (array_key_exists($attr, $diff)) {
+                $result[] = [
+                    'attrib' => $attr,
+                    'data' => $diff[$attr],
+                ];
+            }
+        }
+
+        return $result;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function transformQuery(?array $query = null)
+    {
+        if ($this->filter_all !== null && empty($query)) {
+            return QueryTransformer::transform($this->getFilterAll());
+        }
+        if (!empty($query)) {
+            if ($this->filter_all === null) {
+                return QueryTransformer::transform($query);
+            }
+
+            return QueryTransformer::transform([
+                '$and' => [
+                    $this->getFilterAll(),
+                    $query,
+                ],
+            ]);
+        }
+
+        return [null, []];
+    }
+
+    public function change(AttributeMapInterface $map, array $diff, array $object, EndpointObjectInterface $endpoint_object, bool $simulate = false): ?string
+    {
+        if ($diff === []) {
+            $this->logger->debug('object on endpoint is already up2date', [
+                'category' => get_class($this),
+            ]);
+
+            return null;
+        }
```

Comment:
There is no need for this check. The engine does not call change() if the diff is empty.